templates/README.md

```diff
@@ -0,0 +1,254 @@
+
+The UI can then operate on top of template parameters and/or the Template.VM.spec DOM directly. There is an openshift [REST endpoint for instantiating Templates](https://docs.openshift.com/container-platform/3.7/rest_api/examples.html#template-instantiation) and the UI might be able to use it to create an in memory instance of VM it will then work with before posting it (kubernetes dry-run issues [https://github.com/kubernetes/kubernetes/issues/11488](#11488) and [https://github.com/kubernetes/kubernetes/issues/63559](#63559)).
+
+## Example file
+
+### template.yaml
+
+```YAML
+# vim: sts=2 sw=2 et
+apiVersion: v1
+kind: Template # Openshift kind of template or something "better"
+metadata:
+  name: windows-10
+  annotations:
+    openshift.io/display-name: "Generic Microsoft Windows 7+ template"
+    description: "Basic windows template long description with details"
+    openshift.io/long-description: >-
+      Long description of the template
+    openshift.io/provider-display-name: "Red Hat, Inc."
+    openshift.io/documentation-url: "https://cnv.io/..."
+    openshift.io/support-url: "https://access.redhat.com"
+    iconClass: icon-windows    
+
+    # This set of annotations is meant as a hint only, it is
+    # not going to be processed by the stock openshift templating
+    # mechanism. CNV UI can and should use it though.
+    # The information encoded here informs the tooling (UI)
+    # about names of devices that are to be used as templates
+    # for adding additional disk, networks, volumes and such
+    defaults.template.cnv.io/disk: default-disk
+    defaults.template.cnv.io/volume: default-volume
+    defaults.template.cnv.io/network: default-network
+
+    # Replacements useful for finishing the transformation
+    # This is an extension we could use for specifying common
+    # fields for all disks/nics. Requires extra implementation
+    # that will do the find and replace using a syntax documented
+    # here: https://kubernetes.io/docs/reference/kubectl/jsonpath/
+    # The jsonpath root is the objects: element of the template
+    # XXX There is currently an issue with annotations not supporting
+    #     list values
+    template.cnv.io/replace:
+    - path: "/VM/disks/*/type"
+      value: virtio
+
+    # Another extension useful for CNV aware tooling that allows
+    # expressing additional validation rules for this template
+    # The jsonpath root is the objects: element of the template
+    # XXX There is currently an issue with annotations not supporting
+    #     list values
+    template.cnv.io/validate:
+    - path: "/VM/spec/domain/guest/memory"
+      min: 3
+      max: 4
+
+    # Another extension for hinting at which elements should be
+    # considered editable. The content is a line separated
+    # list of jsonpath selectors.
+    # The jsonpath root is the objects: element of the template
+    template.cnv.io/editable: |
```

Comment:
I don't think this is actually needed. UI knows which parts of the spec it can edit and I don't think it makes sense to configure some parts where we want to disallow it.

Reply:
Fabian indeed wanted to be able to limit the editable areas, there is a use case for that. Think about GCE or AWS where you only select the template (t2.small), but are not allowed to customize it.